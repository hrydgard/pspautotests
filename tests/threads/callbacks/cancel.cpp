#include "shared.h"

int cbFunc(int arg1, int arg2, void *arg) {
	checkpoint(" * cbFunc hit: %08x, %08x, %08x", arg1, arg2, arg);
	return 0;
}

inline void testCancel(const char *title, SceUID cb) {
	int result = sceKernelCancelCallback(cb);
	if (result == 0) {
		checkpoint(NULL);
		schedf("%s: OK ", title);
		schedfCallback(cb);
	} else {
		checkpoint("%s: Failed (%08x)", title, result);
	}
}

extern "C" int main(int argc, char *argv[]) {
	Callback cb("cancel", &cbFunc, NULL);

	checkpointNext("Objects:");
	testCancel("  Normal", cb);
	testCancel("  NULL", 0);
	testCancel("  Invalid", 0xDEADBEEF);
	cb.Delete();
	testCancel("  Deleted", cb);

	cb.Create("notify", &cbFunc, NULL);
	checkpointNext("After notify:");
	sceKernelNotifyCallback(cb, 0x123);
	testCancel("  1 pending", cb);

	checkpoint("sceKernelDelayThreadCB: %08x", sceKernelDelayThreadCB(1000));

	checkpointNext("Different thread:");
	{
		CallbackSleeper waiter1("better priority sleeping thread", 0x10);
		CallbackSleeper waiter2("worse priority sleeping thread", 0x30);
		sceKernelDelayThread(1000);
		sceKernelNotifyCallback(waiter1.callbackID(), 0x1337);
		sceKernelNotifyCallback(waiter2.callbackID(), 0x1337);
		sceKernelCancelCallback(waiter1.callbackID());
		sceKernelCancelCallback(waiter2.callbackID());
		sceKernelDelayThread(1000);
	}

	return 0;
}