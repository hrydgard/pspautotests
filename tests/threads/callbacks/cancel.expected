[x] Objects:
[x]   Normal: OK Callback: OK (size=56,name=cancel,thread=1,callback=1,common=00000000,notifyCount=00000000,notifyArg=0)
[x]   NULL: Failed (800201a1)
[x]   Invalid: Failed (800201a1)
[x]   Deleted: Failed (800201a1)

[x] After notify:
[x]   1 pending: OK Callback: OK (size=56,name=notify,thread=1,callback=1,common=00000000,notifyCount=00000000,notifyArg=0)
[r] sceKernelDelayThreadCB: 00000000

[x] Different thread:
[x]   Beginning sleep on better priority sleeping thread
[x]   ** started better priority sleeping thread
[x]   ** started worse priority sleeping thread
[r]   Beginning sleep on worse priority sleeping thread
[r]   Callback hit on better priority sleeping thread: 00000001, 00001337, returning 00000000
[r]   ** stopped worse priority sleeping thread
[r]   ** stopped better priority sleeping thread
